File: src/test/java/org/storm/engine/example/ParticleTestState.java
package org.storm.engine.example;

import org.storm.core.input.action.ActionManager;
import org.storm.core.ui.Resolutions;
import org.storm.engine.KeyActionConstants;
import org.storm.engine.request.Request;
import org.storm.engine.request.RequestQueue;
import org.storm.engine.request.types.AddForceRequest;
import org.storm.engine.request.types.TogglePhysicsRequest;
import org.storm.physics.entity.Entity;
import org.storm.physics.enums.Direction;
import org.storm.physics.math.geometry.Point;
import org.storm.physics.math.geometry.shapes.Circle;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ThreadLocalRandom;

public class ParticleTestState extends SwitchableState {

  public ParticleTestState() {
    super();
    this.entities.add(new ImmovableRectEntity(0, 0, this.unitConvertor.toUnits(Resolutions.SD.getWidth()), this.unitConvertor.toUnits(5)));
    this.entities.add(new ImmovableRectEntity(0, 0, this.unitConvertor.toUnits(5), this.unitConvertor.toUnits(Resolutions.SD.getHeight())));
    this.entities.add(new ImmovableRectEntity(this.unitConvertor.toUnits(Resolutions.SD.getWidth() - 5), 0, this.unitConvertor.toUnits(5), this.unitConvertor.toUnits(Resolutions.SD.getHeight())));
    this.entities.add(new ImmovableRectEntity(0, this.unitConvertor.toUnits(Resolutions.SD.getHeight() - 5), this.unitConvertor.toUnits(Resolutions.SD.getWidth()), this.unitConvertor.toUnits(5)));
  }

  @Override
  public void preload(RequestQueue requestQueue) {
    List<Request> initialRequests = new ArrayList<>();
    Set<Point> usedPoints = new HashSet<>();
    for (int i = 0; i < 1000; i++) {
      Point topLeft = new Point(ThreadLocalRandom.current().nextInt(10, (int) this.unitConvertor.toUnits(Resolutions.SD.getWidth() - 10)), ThreadLocalRandom.current().nextInt(10, (int) this.unitConvertor.toUnits(Resolutions.SD.getHeight() - 10)));
      while (usedPoints.contains(topLeft)) {
        topLeft = new Point(ThreadLocalRandom.current().nextInt(10, (int) this.unitConvertor.toUnits(Resolutions.SD.getWidth() - 10)), ThreadLocalRandom.current().nextInt(10, (int) this.unitConvertor.toUnits(Resolutions.SD.getHeight() - 10)));
      }
      usedPoints.add(topLeft);

      Entity e = new EntityImpl(new Circle(topLeft.getX(), topLeft.getY(), this.unitConvertor.toUnits(2)), this.unitConvertor.toUnits(2), 0.5, 1);
      this.entities.add(e);
      initialRequests.add(new AddForceRequest(e, Direction.random().getVector().scale(this.unitConvertor.toUnits(2)), 0.1));
    }

    requestQueue.submit(initialRequests);
  }

  @Override
  public void process(ActionManager actionManager, RequestQueue requestQueue) {
    super.process(actionManager, requestQueue);
    if (actionManager.isPerforming(KeyActionConstants.SPACE)) {
      requestQueue.submit(new TogglePhysicsRequest(false));
    }
  }

}

File: components/engine/src/test/kotlin/org/storm/engine/example/ParticleTestState.kt
package org.storm.engine.example

import org.storm.core.context.Context
import org.storm.core.context.RESOLUTION
import org.storm.core.context.RESOLUTION_IN_UNITS
import org.storm.core.extensions.units
import org.storm.core.input.action.ActionState
import org.storm.core.graphics.geometry.Point
import org.storm.core.sound.SoundManager
import org.storm.engine.KeyActionConstants
import org.storm.engine.context.REQUEST_QUEUE
import org.storm.engine.request.types.TogglePhysicsRequest
import org.storm.physics.PhysicsEngine
import org.storm.physics.collision.Collider
import org.storm.physics.enums.Direction
import org.storm.physics.math.geometry.shapes.AABB
import org.storm.physics.math.geometry.shapes.Circle
import java.util.concurrent.ThreadLocalRandom

class ParticleTestState : SwitchableState() {

    private val resolution = Context.RESOLUTION_IN_UNITS

    override val colliders: Set<Collider> = run {
        val colliders = mutableSetOf(
            Collider(
                mutableMapOf(
                    "platformTop" to AABB(
                        0.0,
                        0.0,
                        resolution.width,
                        5.0.units
                    ),
                    "platformLeft" to AABB(
                        0.0,
                        0.0,
                        5.0.units,
                        resolution.height
                    ),
                    "platformRight" to AABB(
                        resolution.width - 5.units,
                        0.0,
                        5.0.units,
                        resolution.height
                    ),
                    "platformBottom" to AABB(
                        0.0,
                        resolution.height - 5.units,
                        resolution.width,
                        5.0.units
                    )
                ),
                Double.POSITIVE_INFINITY,
                1.0
            )
        )

        val usedPoints: MutableSet<Point> = mutableSetOf()

        for (i in 0..400) {
            var topLeft = Point(
                ThreadLocalRandom.current().nextInt(5, (Context.RESOLUTION.width - 5).toInt()).units,
                ThreadLocalRandom.current().nextInt(5, (Context.RESOLUTION.height - 5).toInt()).units
            )

            while (usedPoints.contains(topLeft)) {
                topLeft = Point(
                    ThreadLocalRandom.current().nextInt(5, (Context.RESOLUTION.width - 5).toInt()).units,
                    ThreadLocalRandom.current().nextInt(5, (Context.RESOLUTION.height - 5).toInt()).units
                )
            }

            usedPoints.add(topLeft)

            val c = Collider(
                Circle(topLeft.x, topLeft.y, 2.0.units),
                5.0.units,
                1.0
            )

            colliders.add(c)
        }

        colliders
    }

    override suspend fun onRegister(physicsEngine: PhysicsEngine, soundManager: SoundManager) {
        colliders.forEach {
            physicsEngine.applyForce(Direction.random().vector.scale(2.0.units), it, 0.1)
        }
    }

    override suspend fun process(actionState: ActionState) {
        super.process(actionState)

        if (actionState.isFirstActivation(KeyActionConstants.SPACE)) {
            Context.REQUEST_QUEUE.submit(TogglePhysicsRequest())
        }
    }

}
